finetrainers/ltx_video/ltx_video_lora.py

```diff
@@ -114,19 +127,52 @@ def prepare_latents(
     device: Optional[torch.device] = None,
     dtype: Optional[torch.dtype] = None,
     generator: Optional[torch.Generator] = None,
+    precompute: bool = False,
 ) -> torch.Tensor:
     device = device or vae.device
-    dtype = dtype or vae.dtype
 
     if image_or_video.ndim == 4:
         image_or_video = image_or_video.unsqueeze(2)
     assert image_or_video.ndim == 5, f"Expected 5D tensor, got {image_or_video.ndim}D tensor"
 
-    image_or_video = image_or_video.to(device=device, dtype=dtype)
+    image_or_video = image_or_video.to(device=device, dtype=vae.dtype)
     image_or_video = image_or_video.permute(0, 2, 1, 3, 4).contiguous()  # [B, C, F, H, W] -> [B, F, C, H, W]
-    latents = vae.encode(image_or_video).latent_dist.sample(generator=generator)
-    _, _, num_frames, height, width = latents.shape
-    latents = _normalize_latents(latents, vae.latents_mean, vae.latents_std)
+    if not precompute:
+        latents = vae.encode(image_or_video).latent_dist.sample(generator=generator)
+        latents = latents.to(dtype=dtype)
+        _, _, num_frames, height, width = latents.shape
+        latents = _normalize_latents(latents, vae.latents_mean, vae.latents_std)
+        latents = _pack_latents(latents, patch_size, patch_size_t)
+        return {"latents": latents, "num_frames": num_frames, "height": height, "width": width}
+    else:
+        if vae.use_slicing and image_or_video.shape[0] > 1:
+            encoded_slices = [vae._encode(x_slice) for x_slice in image_or_video.split(1)]
+            h = torch.cat(encoded_slices)
+        else:
+            h = vae._encode(image_or_video)
+        _, _, num_frames, height, width = h.shape
+        # TODO(aryan): this is very very very stupid, but anything to make it work for now. refactor and design better later
```

Comment:
Yes sorry I should have been more clear.

So, we are returning vae latents_mean and latents_std here because we need it for latent normalization in `post_latent_preparation`. But that method is not called until training starts and a DiagonalGaussianDistribution is created. In our precomputation, we also end up saving the latents_mean and latents_std per latent file, which can quickly add up to use a lot of disk storage for bigger precomputation runs (for lora, it doesn't matter too much because it's only about < 100 videos most of the time). Ideally, we should only save this once but since I didn't find time to look into implementing that yet, I just let it be the way it is

Reply:
We can always obtain `latents_mean` and `latents_std` from the VAE config file, right? And loading a config file is very easy with `load_config`. Should be a relatively easy change.

Reply:
It's not easy in this case, which is why I implemented it the way it is 😅 The mean and std are per channel tensors of shape (128,) part of the state dict. Maybe I can just copy the values and paste here as a list. Would that work?

Reply:
I can look into it then. Not all the VAEs would have std and mean so, that will also need to be conditioned accordingly. Leave [this comment](https://github.com/a-r-r-o-w/finetrainers/pull/129#discussion_r1895532692) as a note, then.

Reply:
I expected [this VAE config](https://huggingface.co/Lightricks/LTX-Video/blob/main/vae/config.json) to have `latents_std` and `latents_mean`, following what we do in others. Why does it not have things? 

> Is it okay if I copy the values from the state dict and directly paste into the file?

Yeah okay with a note.